common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -827,13 +833,16 @@ void resetChunkTracker() {
     void updateChunkInfo(long chunkOffset, int mapIndex) throws IOException {
       long idxStartPos = -1;
       try {
-        // update the chunk tracker to meta file before index file
-        writeChunkTracker(mapIndex);
         idxStartPos = indexFile.getFilePointer();
         logger.trace("{} shuffleId {} reduceId {} updated index current {} updated {}",
           appShuffleId.appId, appShuffleId.shuffleId, reduceId, this.lastChunkOffset,
           chunkOffset);
-        indexFile.writeLong(chunkOffset);
+        indexFile.write(Longs.toByteArray(chunkOffset));
+        // Chunk bitmap should be written to the meta file after the index file because if there are
+        // any exceptions during writing the offset to the index file, meta file should not be
+        // updated. If the update to the index file is successful but the update to meta file isn't
+        // then the index file position is reset in the catch clause.
+        writeChunkTracker(mapIndex);
```

Comment:
Given the focus here is error handling, how are we handling a `seek` failure in `catch` block ?

Reply:
This is a good point. I have overlooked IOExceptions from the seeks. Any IOException from the seek would not guarantee that the position is updated. So any more updates to the file will not overwrite the corrupt data which means that the files are corrupted.
One way of handling this I think is maintain the expected length of both index and meta files in AppShufflePartitionInfo instead of relying on the FilePointer. Updates to these files would be similar to writing to the merged data file as below:

```
     if (partitionInfo.isEncounteredFailure()) {
          long updatedPos = partitionInfo.getPosition() + length;
          logger.debug(
            "{} shuffleId {} reduceId {} encountered failure current pos {} updated pos {}",
            partitionInfo.appShuffleId.appId, partitionInfo.appShuffleId.shuffleId,
            partitionInfo.reduceId, partitionInfo.getPosition(), updatedPos);
          length += partitionInfo.dataChannel.write(buf, updatedPos);
        } else {
          length += partitionInfo.dataChannel.write(buf);
        }
```
Please let me know if there are any other suggestions?

Reply:
On a second thought, if there was an exception writing and then again an exception while seek, is there a possibility that any further updates to these files will succeed?
Usually these exceptions wouldn't indicate any transient error with the FileSystem. 
Wouldn't it be better to just stop merging any more blocks of this shuffle partition?

Reply:
I don't think we should assume that.
We could have taken an approach where we give up on writing merged shuffle data file as well.
The issue I see here is that we are mixing block level exception with chunk level exception.
Right now, both will throw exceptions triggering the `onFailure` handling.
However, the current `onFailure` handling is meant for block level only.
I think we should catch the chunk level IOExceptions inside `onComplete`, setting certain flags to make the `onFailure` handling logic know about whether we encountered a block level failure or a chunk level failure.
For a chunk level failure, we shouldn't overwrite the previous block, we should effectively only delay the closure of the chunk.

Reply:
> I don't think we should assume that.
We could have taken an approach where we give up on writing merged shuffle data file as well.

Here the shuffle service is writing to a local file. It's not a network filesystem/hdfs where because of network there could be transient exceptions. If during updating the local index/meta file the shuffle service sees an IOException and then there is an IOException again while seeking to a position, chances that the updates to the same file will fail are very high. If index/meta files are not successfully updated then those merged blocks cannot be served to the executors when they read/fetch it.
If most of the time `IOExceptions` while writing to local filesystem can only be because of issues like file permission, disk full, etc, I don't see the advantage of still trying to keep merging blocks to the shuffle partition.
It might be better for the server to reply to the executor to stop pushing in this case. Otherwise, they keep pushing more blocks and in most of the cases they will fail as well.

Reply:
Stopping merging blocks only after 2 consecutive IOExceptions seems to be giving up too early.
If there's a need to stop early due to too many disk failures, I think that logic should be on the client side inside `ErrorHandler`.
For the server side, it should just handle the received merge requests as best as it can.
Also, index/meta file writing failure should be recoverable just like how we recover data file writing failures.
We just need to handle the data file and index/meta file IOExceptions separately.

Reply:
> I think we should catch the chunk level IOExceptions inside onComplete, setting certain flags to make the onFailure handling logic know about whether we encountered a block level failure or a chunk level failure.
For a chunk level failure, we shouldn't overwrite the previous block, we should effectively only delay the closure of the chunk.

The closure of chunk write now depends on the minimum size of the chunk. If we keep growing the chunk then eventually it will be very large and the executors will get killed  fetching them because they will exceed the physical memory limit. This was the main reason to serve data in manageable-sized chunks.

> Stopping merging blocks only after 2 consecutive IOExceptions seems to be giving up too early.

With local file systems the IOExceptions that I can think of would be due to disk failures, file corruption, permission and these will not go away when retrying to write to these. Are there any other scenarios that I am missing?

> If there's a need to stop early due to too many disk failures, I think that logic should be on the client side inside ErrorHandler.

Sure, however the server here can respond with a similar runtime exception that it can't update merge file like the `TooLate` runtime exception and the error handle for the push side can stop pushing the data.

Reply:
There are temporary issues as well.
For example, a user's job might be filling up the yarn local dir with its temporary data, which would then lead to IOExceptions for merge writes.
When this issue occurs, it could trigger a clean up script which purges the temporary data from that job to recover the disk.
In this scenario, the IOException could be temporary.

For the too large chunk issue, it would only be too large if we have many consecutive IOExceptions.
Once recovered, the following chunks would go back to the normal size.
Only 2 consecutive IOExceptions within the same chunk is far from causing the potential too large chunk issue.

If we want to have a reasonable policy to stop early due to too many IOExceptions, I think that policy should require more than 2 consecutive IOExceptions during chunk metadata write to trigger.
Note that, the same issue could affect writing data file as well.
If we want to address this issue holistically, it would require some state tracking.
If so, I'd rather track it on the client side instead of the server side.

Reply:
@Victsm The case that you described how often does it happen? How does this cleanup script differentiates shuffle data from temporary user data and how quickly does it remove this data. This has an assumption that cluster is setup in a certain way.
Would you agree that the more common cases are the non-recoverable disk failures, permission issues, file corruption etc?
If yes, then there will be continuous IOExceptions while updating these index/meta file. So if we just let the chunk grow when the index/meta file couldn't be updated, in most of the cases the executor will fail to fetch these.

In the non-push shuffle mechanism, if writing to index file fails does the executor try multiple times to write to the same file locally? I understand that in the push-mechanism, we are updating the index/meta in a continuous manner but these are also just local files.

> I think that policy should require more than 2 consecutive IOExceptions during chunk metadata write to trigger.
Note that, the same issue could affect writing data file as well.

Just a single IOException while updating the index/meta file should be enough to indicate the client to stop pushing. Yes, it affects writing the data file as well. The reason I am not bringing that up because in this change we are focusing on the exceptions while updating metadata. If others think that we should address the failures during writes to data file as well in this followup fix, we can address it. 

> If we want to address this issue holistically, it would require some state tracking.
If so, I'd rather track it on the client side instead of the server side.

We are already tracking the state of shuffle partition in `AppShufflePartitionInfo`. Adding few other flags to it shouldn't be an issue.

Would like to know the thoughts of other folks as well?
@mridulm @tgravescs @attilapiros @Ngone51

Reply:
According to our operation experiences, the temporary issue I described is more common than the non-recoverable issues you described.

Reply:
Also, the script I described is just one potential way to mitigate the issue faster.
That's temporary data which will be deleted by YARN at the end of the job anyway, so the cluster already has ways to self-recover from that type of issues.

Reply:
Giving up on the merge write just after one IOException is too expensive.
We could have merged much more blocks if such an issue is only temporary.
I think this is going against our original design to make the shuffle partition merge write tolerant of duplication, collision, and failure.

Reply:
Tracking IO exceptions sounds a little bit complex to me...I'd prefer to stop merging in case of IO exception for the seek.

Besides, I'm thinking it would be good if we could dynamically change the merger location when such IO exception happens.

Reply:
Had a discussion with @Victsm @mridulm yesterday. This is the approach currently we are thinking of:
1. Verify if `seek` is not recoverable. If it is not then let the clients know to stop pushing and stop merging this partition.
2. Have a threshold on number of `IOExceptions` from writes and when this threshold is reached for a single partition, inform the client to stop pushing and stop merging this partition.
3. When  the update to metadata fails, not propagate this exception back to client so that they push the block again. The size of the current chunk may grow but with (2) in place it will still be of a manageable size.

These changes will not be that complex.

Reply:
> Besides, I'm thinking it would be good if we could dynamically change the merger location when such IO exception happens.

Which one of these do you mean?
1.  The executor should start pushing to a different server for the current shuffle that is going on.
2.  Do you mean that for a future shuffle, the driver should decide the merger locations considering these IOExceptions. 

(1) is currently difficult. The driver decides the specific merger locations before a shuffle stage and all the executors work on the assumption that these are all the shuffle servers that they push the data to. This list of shuffle servers should be consistent across the shuffle servers because if it's not then they will be push the shuffle data belonging to a reducer to different shuffle servers. This is the code I am talking about- [code](https://github.com/linkedin/spark/blob/002cb69e8ddf49bbe114744b84c46e0fd452d852/core/src/main/scala/org/apache/spark/shuffle/ShuffleBlockPusher.scala#L360).

(2) This could be a future improvement where we also include the metrics for merge reported back to driver when the driver triggers the finalize. The driver could use these metrics when deciding which shuffle server to use for the next push.

Reply:
I mean (1). I know it's hard for the current framework. It requires more refactor if we want to do so. 

(2) sounds good to me. But to clarify, it only benefits the following push but doesn't help ease the problem(the IO exception) in the current push.

Reply:
For dynamically changing merger locations for the current push, if the non-recoverable IOException happen after majority of the blocks have already been merged, we won't gain much by pushing the remaining blocks in this shuffle targeted to the failed shuffle service to a new location. Probably need to check the merge ratio before deciding whether to do so or not. The current framework is designed around the best-effort nature of the block push operation, which simplifies the handling of these failure scenarios.

Reply:
Openjdk `seek` seems to use `lseek64`
```
// move file pointer to the specified offset
jlong os::seek_to_file_offset(int fd, jlong offset) {
  return (jlong)::lseek64(fd, (off64_t)offset, SEEK_SET);
}
```
`lseek64` is just a variation of `lseek` that uses 64 bit file offset.
Below are the errors from https://linux.die.net/man/2/lseek
```
EBADF
fd is not an open file descriptor.
EINVAL
whence is not valid. Or: the resulting file offset would be negative, or beyond the end of a seekable device.
EOVERFLOW
The resulting file offset cannot be represented in an off_t.
ESPIPE
fd is associated with a pipe, socket, or FIFO.
ENXIO
whence is SEEK_DATA or SEEK_HOLE, and the current file offset is beyond the end of the file.
```
In our case, we always seek to a previous point in the file and none of these errors would be recoverable in that case.
So any exceptions during seek should not be recoverable.
@mridulm @Victsm

Reply:
Thanks for confirming @otterc.
We have two paths here:
a) Ignore seek exception as unexpected.
b) Abort merge, assuming EBADF/etc - as it is unrecoverable error.

I am partial towards (b) as we cannot recover with seek exception. Thoughts ?

Reply:
Just want to reiterate the change I am working on:
- Since seek is not recoverable, we let the clients know to stop pushing for a particular shuffle partition.
- Have a threshold on number of IOExceptions from writes and when this threshold is reached for a single partition, inform the client to stop pushing and stop merging this partition.
- When the update to metadata fails, not propagate this exception back to client so that they push the block again. The size of the current chunk may grow but with (2) in place it will still be of a manageable size.

These changes will impact https://github.com/apache/spark/pull/30312. Right now the client stops pushing for the entire shuffle not a particular shuffle partition. The support needs to be extended to stop pushing for a particular shuffle partition.

Reply:
Could you file a JIRA ticket for the corresponding work for the client?

Reply:
I created this Jira:
https://issues.apache.org/jira/browse/SPARK-33665